mos-platform/common/include/fixed_point.h

```diff
@@ -0,0 +1,301 @@
+
+template<uint8_t ISize, uint8_t FSize, bool Signed>
+class Value {
+protected:
+
+  // Used to calculate the bitmasks for both the integer part and fractional part
+  template <typename R>
+  static constexpr R BitMask(unsigned int const onecount) {
+    return static_cast<R>(-(onecount != 0))
+      & (static_cast<R>(-1) >> ((sizeof(R) * CHAR_BIT) - onecount));
+  }
+
+  // Size of the underlying storage for the different views of the data.
+  // The total combined storage will be used for representing the actual value.
+  using IntType = std::conditional_t<Signed, 
+                                     signed _BitInt(((ISize + 7) / 8) * 8),
+                                     unsigned _BitInt(((ISize + 7) / 8) * 8)>;
+  using FracType = std::conditional_t<Signed,
+                                      signed _BitInt(((FSize + 7) / 8) * 8),
+                                      unsigned _BitInt(((FSize + 7) / 8) * 8)>;
+  using StorageType = std::conditional_t<Signed,
+                                         signed _BitInt((((ISize + FSize) + 7) / 8) * 8),
+                                         unsigned _BitInt((((ISize + FSize) + 7) / 8) * 8)>;
+  StorageType val;
+  static constexpr unsigned long INT_MASK = BitMask<unsigned long>(ISize) << FSize;
+  static constexpr unsigned long FRAC_MASK = BitMask<unsigned long>(FSize);
+
+public:
+
+  // Constructors
+
+  /// Common constructor for creating a fixed point value from an integer value
+  [[clang::always_inline]] constexpr explicit Value(IntType i) {
+    SetI(i); 
+  }
+  
+  /// Copy constructor for a fixed point value with different sizes of Int/Frac
+  template<uint8_t OI, uint8_t OF, bool S>
+  [[clang::always_inline]] constexpr Value(const Value<OI, OF, S>& o) {
+    val = Value<ISize, FSize>(o.AsI(), o.AsF()).Get();
+  }
+  /// Copy constructor for fixed point values with the same size
+  [[clang::always_inline]] constexpr Value(const Value& o) : val(o.val) {}
+  
+  /// Constructor for setting both the integral and fractional part
+  [[clang::always_inline]] constexpr explicit Value(IntType i, FracType f) {
+    SetI(i);
+    SetF(f);
+  }
+  /// Constructor to convert floating point values into fixed point
+  [[clang::always_inline]] constexpr Value(long double f) {
+    Set(f * (1 << FSize));
+  }
+
+  // Direct value accessor and setter methods
+
+  /// Returns just the integral portion
+  [[clang::always_inline]] constexpr IntType AsI() const { return (val & INT_MASK) >> FSize; }
+  /// Returns just the fractional portion
+  [[clang::always_inline]] constexpr FracType AsF() const { return (val & FRAC_MASK); }
+  /// Returns the entire value
+  [[clang::always_inline]] constexpr StorageType Get() const { return val; }
+  /// Update just the integral portion
+  [[clang::always_inline]] constexpr void SetI(IntType v) { 
+    val = (val & ~INT_MASK) | static_cast<StorageType>((v) & (INT_MASK >> FSize)) << FSize;
```

Comment:
This kind of logic seems a lot easier to express using bit fields in a struct; does that also produce sufficient codegen quality?